web/src/js/ducks/flows/index.ts

```diff
@@ -0,0 +1,366 @@
+export const FLOWS_UPDATE = createAction<{
+    flow: Flow;
+    matching_filters: Partial<{ [key in FilterName]: boolean }>;
+}>("FLOWS_UPDATE");
+export const FLOWS_REMOVE = createAction<string>("FLOWS_REMOVE");
+export const FLOWS_RECEIVE = createAction<Flow[]>("FLOWS_RECEIVE");
+export const FLOWS_FILTER_UPDATE = createAction<{
+    name: FilterName;
+    matching_flow_ids: string[] | null;
+}>("FLOWS_FILTER_UPDATE");
+
+export const setSort = createAction<{
+    column?: keyof typeof sortFunctions;
+    desc: boolean;
+}>("flows/sort");
+export const select = createAction<Flow[]>("flows/select");
+
+export interface FlowsState {
+    list: Flow[];
+    _listIndex: Map<string, number>;
+    byId: Map<string, Flow>;
+    view: Flow[];
+    _viewIndex: Map<string, number>;
+
+    sort: { column?: keyof typeof sortFunctions; desc: boolean };
+    selected: Flow[];
+    selectedIds: Set<string>;
+    highlighted: Set<string>;
+}
+
+export const defaultState: FlowsState = {
+    list: [],
+    _listIndex: new Map(),
+    byId: new Map(),
+    view: [],
+    _viewIndex: new Map(),
+
+    sort: { column: undefined, desc: false },
+    selected: [],
+    selectedIds: new Set(),
+    highlighted: new Set(),
+};
+
+// This is a manual reducer as RTK's createSlice always uses Immer, which is orders of magnitude slower.
+// In benchmarking we found no major difference between Map/Set and objects, so we use objects
+// for redux compatibility.
+export default function flowsReducer(
+    state = defaultState,
+    action: UnknownAction,
+): FlowsState {
+    if (FLOWS_RECEIVE.match(action)) {
+        const { sort } = state;
+        const list = action.payload;
+        const _listIndex = buildIndex(list);
+        const byId = new Map(list.map((f) => [f.id, f]));
+        // No filter information yet, we expect that to come immediately after.
+        const view = list.toSorted(makeSort(sort));
+        const _viewIndex = buildIndex(view);
+        const selected = state.selected
+            .map((flow) => byId.get(flow.id))
+            .filter((f) => f !== undefined);
+        const selectedIds = buildLookup(selected);
+        const highlighted = new Set<string>();
+
+        return {
+            list,
+            _listIndex,
+            byId,
+            view,
+            _viewIndex,
+            sort,
+            selected,
+            selectedIds,
+            highlighted,
+        };
+    } else if (FLOWS_ADD.match(action)) {
+        const { flow, matching_filters } = action.payload;
+        const { sort, selected, selectedIds } = state;
+        let { view, _viewIndex, highlighted } = state;
+        // Update list
+        const _listIndex = new Map(state._listIndex);
+        _listIndex.set(flow.id, state.list.length);
+        const list = [...state.list, flow];
+        const byId = new Map(state.byId);
+        byId.set(flow.id, flow);
+        // Update view
+        if (matching_filters[FilterName.Search] !== false) {
```

Comment:
for readability I would stick to `=== true` like below or simply `if (matching_filters[FilterName.Search])`

Reply:
https://github.com/mitmproxy/mitmproxy/pull/7704/commits/d28bd5ec4a7de1baffcf5b3a1c0b58e69de7b8f6 :)